source/MRViewer/MRRibbonNotification.cpp

```diff
@@ -25,28 +25,227 @@ constexpr std::array< std::pair<const char*, ImU32>, int( MR::NotificationType::
     std::pair<const char*, ImU32> { "\xef\x8b\xb2", ::MR::Color( 255, 146, 0 ).getUInt32() }
 };
 
+const float cWindowRounding = 4.f;
+const float cWIndowSpacing = 20.f;
+
 }
 
 namespace MR
 {
 
 void RibbonNotifier::pushNotification( const RibbonNotification& notification )
 {
-    if ( !notifications_.empty() && notifications_.front().notification == notification )
+    if ( !historyMode_ )
+        addNotification_( notifications_, notification );
+    addNotification_( notificationsHistory_, notification );
+
+    highestNotification_ = NotificationType::Count;
+    for ( const auto& nwt : notificationsHistory_ )
     {
-        notifications_.front().sameCounter++;
-        notifications_.front().timer = 0.0f;
+        if ( int( nwt.notification.type ) < int( highestNotification_ ) )
+            highestNotification_ = nwt.notification.type;
     }
+    
+    requestClosestRedraw_();
+}
+
+void RibbonNotifier::draw( float scaling )
+{
+    if ( historyMode_ )
+        drawHistory_( scaling );
     else
+        drawNotifications_( scaling );
+    drawHistoryButton_( scaling );
+    filterInvalid_( -1 );
+}
+
+void RibbonNotifier::drawHistoryButton_( float scaling )
+{
+    if ( notificationsHistory_.empty() )
+        return;
+
+    float notificationsPosX = cWIndowSpacing * scaling;
+    if ( auto menu = getViewerInstance().getMenuPluginAs<RibbonMenu>() )
+        notificationsPosX += float( menu->getSceneSize().x );
```

Comment:
instead of dynamic casting RibbonMenu mb take it as argument from draw? or set as class field? or better `sceneSize`